HW1/P2/P2.txt

```diff
@@ -1 +1,5 @@
-# Your discussion here
+The two methods utilized in this problem were the default approach of assigning partitions in order and the special approach of randomizing the partitions to distribute the load among the workers. The difficulty in finding a good parallelization balance for this problem came from the fact that it required an uneven amount of concentrated work from a few of the partitions, and therefore made the runtime on some of the workers exponentially greater than on others. Since this is a Mandelbrot problem, the partitions near the middle of the grid were required to perform a lot of calculation as that was where the highest intensity pixels were. The remaining partitions finished relatively quickly as the rest of the graph did not require nearly as much computation. This unbalance may be seen in the first histogram where a few of the partitions are seen to be taking an exorbitant amount of time.
+
+To remedy this issue, a simple randomization of pixels distributed to the partitions was implemented. By shuffling the pixels instead of assigning them as whole chunks, the general work of the high intensity pixels was distributed over several partitions rather than just one or two. The second histogram reveals this new balanced parallelization. It is important to note that this randomization adds shuffle time to the overall runtime. The benefits of the distribution far outweigh the costs in this case, but it is still something to seriously consider. 
+
+One other note is that this randomization, while effective in increasing the spread of the histogram and distributing the load such that the entire parallelization ran faster as a whole, was still quite arbitrary. Though splitting by less arbitrary parameters was not found to be any more effective than simply randomizing, it is potentially possible to take advantage of our knowledge of the problem and its difficulty (concentrated computation on central white pixels) and to devise a scheme to capitalize on this and perform even better. Though I could not think of one for this assignment, this is definitely something to think about for future problems.
```

Comment:
Great writeup - I made a box of the heavy computation in the center and split it evenly but it honestly wasn't much better than the randomized algorithm.